libvast/src/data.cpp

```diff
@@ -335,4 +341,116 @@ bool convert(const data& d, json& j, const type& t) {
   return convert(d, j);
 }
 
+namespace {
+
+data parse(const YAML::Node& node) {
+  switch (node.Type()) {
+    case YAML::NodeType::Undefined:
+    case YAML::NodeType::Null:
+      return data{};
+    case YAML::NodeType::Scalar: {
+      auto str = node.as<std::string>();
+      data result;
+      // Attempt some type inference.
+      if (parsers::boolean(str, result))
+        return result;
+      // Attempt maximum type inference.
+      if (parsers::data(str, result))
+        return result;
+      // Take the input as-is if nothing worked.
+      return str;
+    }
+    case YAML::NodeType::Sequence: {
+      list xs;
+      xs.reserve(node.size());
+      for (size_t i = 0; i < node.size(); ++i)
+        xs.push_back(parse(node[i]));
+      return xs;
+    }
+    case YAML::NodeType::Map: {
+      record xs;
+      xs.reserve(node.size());
+      for (auto& pair : node)
+        xs.emplace(pair.first.as<std::string>(), parse(pair.second));
+      return xs;
+    }
+  }
+  die("unhandled YAML node type");
```

Comment:
That's pretty harsh, why not just throw `YAML::Exception` and fail to parse the file?

Reply:
The idea is that this is more like a developer issue. We enumerated all switch statements above, so we should never reach this code. If so, it's a logic error. In order to make the compiler happy (and not spit out a warning), I plugged in a `[[noreturn]]` function for which we conveniently have `die`.

Reply:
I'm fine with leaving it as-is here, but in general we should be careful with jumping from "programmer error" to "kill the database", especially when there's a perfectly reasonable recovery path.

Related rant: https://lkml.org/lkml/2016/10/4/1